src/corecel/sys/TraceCounter.cuda.cc

```diff
@@ -0,0 +1,89 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/sys/TraceCounter.cuda.cc
+//! \brief Numeric tracing counter
+//---------------------------------------------------------------------------//
+#include "TraceCounter.hh"
+
+#include <cstdint>
+#include <type_traits>
+#include <nvtx3/nvToolsExt.h>
+
+#include "corecel/sys/detail/NvtxUtils.hh"
+
+#include "ScopedProfiling.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Simple performance tracing counter.
+ * \tparam T Arithmetic counter type
+ *
+ * Records a named value at the current timestamp which
+ * can then be displayed on a timeline. Only supported on host
+ */
+template<class T>
+void trace_counter(char const* name, T value)
+{
+    static_assert(std::is_arithmetic_v<T>, "Only support numeric counters");
+    if (use_profiling())
+    {
+        nvtxEventAttributes_t attributes{};
+        attributes.version = NVTX_VERSION;
+        attributes.size = NVTX_EVENT_ATTRIB_STRUCT_SIZE;
+
+        attributes.colorType = NVTX_COLOR_ARGB;
+        attributes.color = 0xFFFF0000;
+
+        attributes.messageType = NVTX_MESSAGE_TYPE_REGISTERED;
+        attributes.message.registered = message_handle_for(name);
+        if constexpr (std::is_same_v<T, std::uint32_t>)
+        {
+            attributes.payloadType = NVTX_PAYLOAD_TYPE_UNSIGNED_INT32;
+            attributes.payload.uiValue = value;
+        }
+        else if constexpr (std::is_same_v<T, std::uint64_t>)
+        {
+            attributes.payloadType = NVTX_PAYLOAD_TYPE_UNSIGNED_INT64;
+            attributes.payload.ullValue = static_cast<uint64_t>(value);
+        }
+        else if constexpr (std::is_same_v<T, float>)
+        {
+            attributes.payloadType = NVTX_PAYLOAD_TYPE_FLOAT;
+            attributes.payload.fValue = value;
+        }
+        else if constexpr (std::is_same_v<T, double>)
+        {
+            attributes.payloadType = NVTX_PAYLOAD_TYPE_DOUBLE;
+            attributes.payload.dValue = value;
+        }
+        else if constexpr (std::is_same_v<T, std::int32_t>)
+        {
+            attributes.payloadType = NVTX_PAYLOAD_TYPE_INT32;
+            attributes.payload.iValue = value;
+        }
+        else if constexpr (std::is_same_v<T, std::int64_t>)
+        {
+            attributes.payloadType = NVTX_PAYLOAD_TYPE_INT64;
+            attributes.payload.llValue = value;
+        }
```

Comment:
Perhaps to improve readability:
```suggestion
#define TC_SET_PAYLOAD(TYPE, NVTX_CONST, MEMBER)            \
    if constexpr (std::is_same_v<T, TYPE>)                  \
    {                                                       \
        attributes.payloadType = NVTX_CONST;                \
        attributes.payload.MEMBER = value; \
    }

    TC_SET_PAYLOAD(std::uint32_t, NVTX_PAYLOAD_TYPE_UNSIGNED_INT32, uiValue)
    else TC_SET_PAYLOAD(std::uint64_t, NVTX_PAYLOAD_TYPE_UNSIGNED_INT64, ullValue)
    else TC_SET_PAYLOAD(float,        NVTX_PAYLOAD_TYPE_FLOAT,          fValue)
    else TC_SET_PAYLOAD(double,       NVTX_PAYLOAD_TYPE_DOUBLE,         dValue)
    else TC_SET_PAYLOAD(std::int32_t, NVTX_PAYLOAD_TYPE_INT32,          iValue)
    else TC_SET_PAYLOAD(std::int64_t, NVTX_PAYLOAD_TYPE_INT64,          llValue)

#undef TC_SET_PAYLOAD
```

Reply:
Or we can avoid macros with this 😅 
```c++
template <typename T>
struct NvtxAttribSetter;

template <>
struct NvtxAttribSetter <std::uint32_t> {
    static void set(nvtxEventAttributes_t& attr, std::uint32_t val) {
        attr.payloadType = NVTX_PAYLOAD_TYPE_UNSIGNED_INT32;
        attr.payload.uiValue = val;
    }
};

template <>
struct NvtxAttribSetter<std::uint64_t> {
    static void set(nvtxEventAttributes_t& attr, std::uint64_t val) {
        attr.payloadType = NVTX_PAYLOAD_TYPE_UNSIGNED_INT64;
        attr.payload.ullValue = val;
    }
};
```

Reply:
Yeah I thought about a struct specialization too (ps you don't even need templating above, you can overload) but for these bunches of changes where we're just trying to compress the text on the screen to *improve* readability, I think macros are appropriate.